lib/requests/request_executor.rb

```diff
@@ -0,0 +1,319 @@
+        start_first_topology_update(urls)
+      elsif (@_without_topology && options["single_node_topology"])
+        @_node_selector = NodeSelector.new(self, options["single_node_topology"])
+      end
+    end
+
+    def self.create(urls, database = nil)
+      return self.new(database, {
+        "without_topology" => false,
+        "first_topology_update_urls" => urls.clone
+      })
+    end
+
+    def self.create_for_single_node(url, database = nil)
+      topology = Topology.new(-1, [ServerNode.new(url, database)])
+
+      return self.new(database, {
+        "without_topology" => true,
+        "single_node_topology" => topology,
+        "topology_etag" => -2
+      })
+    end
+
+    def execute(command)
+      chosen_node = nil
+      chosen_node_index = -1
+
+      await_first_topology_update
+      selector = @_node_selector
+      chosen_node = selector.current_node
+      chosen_node_index = selector.current_node_index
+
+      begin
+        response = execute_command(command, chosen_node)
+      rescue TopologyNodeDownException
+        response = handle_server_down(command, chosen_node, chosen_node_index)
+      end
+
+      response
+    end  
+
+    protected
+    def is_first_topology_update_tries_expired?
+      @_first_topology_updates_tries >= MaxFirstTopologyUpdatesTries
+    end
+
+    def await_first_topology_update()
+      is_fulfilled = false
+      first_topology_update = @_first_topology_update
+
+      if @_without_topology
+        return;
+      end
+
+      @_await_first_topology_lock.synchronize do
+        if first_topology_update.equal?(@_first_topology_update)
+          is_fulfilled = true == first_topology_update;
+
+          if false == first_topology_update
+            start_first_topology_update(@_last_known_urls)
+          end  
+        end
+      end  
+
+      if !is_fulfilled
+        if is_first_topology_update_tries_expired?
+          raise DatabaseLoadFailureException, 'Max topology update tries reached'
+        elsif
+          sleep 0.1
+          await_first_topology_update
+        end    
+      end      
+    end
+
+    def prepare_command(command, server_node)
+      command.create_request(server_node)
+      request = command.to_request_options
+
+      @headers.each do |header, value|
+        request.add_field(header, value)
+      end
+
+      if !@_without_topology
+        request.add_field("Topology-Etag", @_topology_etag)
+      end
+
+      request
+    end
+
+    def execute_command(command, server_node)      
+      if !command.is_a?(RavenCommand)
+        raise InvalidOperationException, 'Not a valid command'
+      end
+
+      if !server_node.is_a?(ServerNode)
+        raise InvalidOperationException, 'Not a valid server node'
+      end
+
+      request = prepare_command(command, server_node)
+      
+      begin
+        response = http_client(server_node).request(request)
+      rescue
+        raise TopologyNodeDownException, "Node #{server_node.url} is down"
```

Comment:
@a-serdukov-mobilunity I was able to pass inner exception info using the third `raise` parameter like this (see `test` method):
```
class TestError < StandardError
end

def imGonnaThrow
    raise StandardError.new('throwing')
end

def seriousStuff
    imGonnaThrow
end

def test
    begin
        seriousStuff
    rescue StandardError => e
        raise TestError, 'TestError', e.backtrace
    end
end

test
```

Getting output like this:
```
gregolsky@gregolsky-pc:/mnt/c/work/experiments/ruby$ ruby test.rb
test.rb:6:in `imGonnaThrow': TestError (TestError)
        from test.rb:10:in `seriousStuff'
        from test.rb:15:in `test'
        from test.rb:22:in `<main>'
gregolsky@gregolsky-pc:/mnt/c/work/experiments/ruby$
```

Not sure if these are the best practices though - examples on the internet do not seem to care about inner exception for some reason?! Also looks like we lose type info on the original exception. In general we need to know what was the first rescued error and bubble that up.